## You're here! 
Welcome to your first competition in the [ITI's AI Pro training program](https://ai.iti.gov.eg/epita/ai-engineer/)! We hope you enjoy and learn as much as we did prepairing this competition.


## Introduction

In the competition, it's required to predict the `Severity` of a car crash given info about the crash, e.g., location.

This is the getting started notebook. Things are kept simple so that it's easier to understand the steps and modify it.

Feel free to `Fork` this notebook and share it with your modifications **OR** use it to create your submissions.

### Prerequisites
You should know how to use python and a little bit of Machine Learning. You can apply the techniques you learned in the training program and submit the new solutions! 

### Checklist
You can participate in this competition the way you perefer. However, I recommend following these steps if this is your first time joining a competition on Kaggle.

* Fork this notebook and run the cells in order.
* Submit this solution.
* Make changes to the data processing step as you see fit.
* Submit the new solutions.

*You can submit up to 5 submissions per day. You can select only one of the submission you make to be considered in the final ranking.*


Don't hesitate to leave a comment or contact me if you have any question!

## Import the libraries

We'll use `pandas` to load and manipulate the data. Other libraries will be imported in the relevant sections.

In [21]:
import pandas as pd
import os

## Exploratory Data Analysis
In this step, one should load the data and analyze it. However, I'll load the data and do minimal analysis. You are encouraged to do thorough analysis!

Let's load the data using `pandas` and have a look at the generated `DataFrame`.

In [22]:
dataset_path = "Users\lenovo\Downloads\test.csv"

df = pd.read_csv('train.csv')

print("The shape of the dataset is {}.\n\n".format(df.shape))

df.head()

The shape of the dataset is (6407, 16).




ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp  
0  2016-03-25 15:13:02  
1  2020-05-05 19:23:00  
2  2016-09-16 19:57:16  
3  2020-03-29 19:48:43  
4  2019-10-09 08:47:00

We've got 6407 examples in the dataset with 14 featues, 1 ID, and the `Severity` of the crash.

By looking at the features and a sample from the data, the features look of numerical and catogerical types. What about some descriptive statistics?

In [23]:
df.drop(columns='ID').describe()

Lat          Lng  Distance(mi)     Severity
count  6407.000000  6407.000000   6407.000000  6407.000000
mean     37.765653  -122.405990      0.135189     2.293429
std       0.032555     0.028275      0.396360     0.521225
min      37.609619  -122.510440      0.000000     1.000000
25%      37.737096  -122.412210      0.000000     2.000000
50%      37.768238  -122.404835      0.000000     2.000000
75%      37.787813  -122.392478      0.041000     3.000000
max      37.825626  -122.349734      6.820000     4.000000

In [24]:
df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp  
0  2016-03-25 15:13:02  
1  2020-05-05 19:23:00  
2  2016-09-16 19:57:16  
3  2020-03-29 19:48:43  
4  2019-10-09 08:47:00

In [25]:
df["Distance(mi)"].value_counts()

0.000    3923
0.010     499
0.037     158
0.420      40
0.208      35
         ... 
0.789       1
0.359       1
1.682       1
0.498       1
0.230       1
Name: Distance(mi), Length: 579, dtype: int64

In [26]:
df['Date'] = pd.to_datetime(df['timestamp']).dt.date
df['Time'] = pd.to_datetime(df['timestamp']).dt.time
df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0     False    False    False       False  False     True    R         2   
1     False    False    False       False  False    False    R         2   
2     False    False    False       False   True    False    R         3   
3      True    False    False       False  False    False    R         1   
4     False    False    False       False  False    False    R         2   

             timestamp        Date      Time  
0  2016-03-25 15:13:02  2016-03-25  15:13:02  
1  2020-05-05 19:23:00  2020-05-05  19:23:00  
2  2016-09-16 19:57:16  2016-09-16  19:57:16  
3  2020-03-29 19:48:43  2020-03-29  19:48:43  
4  2019-10-09 08:47:00  2019-10-09  08:47:00

In [27]:
df["Date"]= pd.to_datetime(df["Date"])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6407 entries, 0 to 6406
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   ID            6407 non-null   int64         
 1   Lat           6407 non-null   float64       
 2   Lng           6407 non-null   float64       
 3   Bump          6407 non-null   bool          
 4   Distance(mi)  6407 non-null   float64       
 5   Crossing      6407 non-null   bool          
 6   Give_Way      6407 non-null   bool          
 7   Junction      6407 non-null   bool          
 8   No_Exit       6407 non-null   bool          
 9   Railway       6407 non-null   bool          
 10  Roundabout    6407 non-null   bool          
 11  Stop          6407 non-null   bool          
 12  Amenity       6407 non-null   bool          
 13  Side          6407 non-null   object        
 14  Severity      6407 non-null   int64         
 15  timestamp     6407 non-null   object  

In [28]:
df['day'] = df['Date'].dt.day
df['month'] = df['Date'].dt.month
df['year'] = df['Date'].dt.year
df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0   0  37.762150 -122.405660  False         0.044     False     False   
1   1  37.719157 -122.448254  False         0.000     False     False   
2   2  37.808498 -122.366852  False         0.000     False     False   
3   3  37.785930 -122.391080  False         0.009     False     False   
4   4  37.719141 -122.448457  False         0.000     False     False   

   Junction  No_Exit  Railway  ...   Stop  Amenity  Side Severity  \
0     False    False    False  ...  False     True     R        2   
1     False    False    False  ...  False    False     R        2   
2     False    False    False  ...   True    False     R        3   
3      True    False    False  ...  False    False     R        1   
4     False    False    False  ...  False    False     R        2   

             timestamp       Date      Time day  month  year  
0  2016-03-25 15:13:02 2016-03-25  15:13:02  25      3  2016  
1  2020-05-05 19:23:00 2020-05-05  19:23:00   5      5  2020  
2  2016-09-16 19:57:16 2016-09-16  19:57:16  16      9  2016  
3  2020-03-29 19:48:43 2020-03-29  19:48:43  29      3  2020  
4  2019-10-09 08:47:00 2019-10-09  08:47:00   9     10  2019  

[5 rows x 21 columns]

The output shows desciptive statistics for the numerical features, `Lat`, `Lng`, `Distance(mi)`, and `Severity`. I'll use the numerical features to demonstrate how to train the model and make submissions. **However you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.**

## Data Splitting

Now it's time to split the dataset for the training step. Typically the dataset is split into 3 subsets, namely, the training, validation and test sets. In our case, the test set is already predefined. So we'll split the "training" set into training and validation sets with 0.8:0.2 ratio. 

*Note: a good way to generate reproducible results is to set the seed to the algorithms that depends on randomization. This is done with the argument `random_state` in the following command* 

In [29]:
from sklearn.model_selection import train_test_split

train_df, val_df = train_test_split(df, test_size=0.2, random_state=42 , stratify = df["Severity"]) # Try adding `stratify` here

X_train = train_df.drop(columns=['ID', 'Severity'])
y_train = train_df['Severity']

X_val = val_df.drop(columns=['ID', 'Severity'])
y_val = val_df['Severity']


As pointed out eariler, I'll use the numerical features to train the classifier. **However, you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.** 

In [30]:
X_train.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
454   37.740299 -122.407848  False         0.000     False     False   
3198  37.709020 -122.467770  False         0.000     False     False   
4652  37.732292 -122.414909  False         0.010     False     False   
1988  37.807710 -122.367640  False         0.037     False     False   
1029  37.709324 -122.467308  False         0.000     False     False   

      Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
454      False    False    False       False  False    False    R   
3198     False    False    False       False  False    False    R   
4652      True    False    False       False  False    False    R   
1988     False    False    False       False  False    False    R   
1029     False    False    False       False  False    False    R   

                timestamp       Date      Time  day  month  year  
454   2020-02-17 12:44:00 2020-02-17  12:44:00   17      2  2020  
3198  2016-11-23 16:11:05 2016-11-23  16:11:05   23     11  2016  
4652  2017-01-02 13:17:06 2017-01-02  13:17:06    2      1  2017  
1988  2018-03-22 15:20:42 2018-03-22  15:20:42   22      3  2018  
1029  2016-08-07 16:16:58 2016-08-07  16:16:58    7      8  2016

In [31]:
X_val.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
5069  37.768948 -122.406372  False           0.0     False     False   
409   37.785835 -122.391518  False           0.0     False     False   
4998  37.710304 -122.465553  False           0.0     False     False   
3249  37.770178 -122.421162  False           0.0     False     False   
4371  37.729990 -122.403729  False           0.0     False     False   

      Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
5069     False    False    False       False  False    False    R   
409       True    False    False       False  False    False    R   
4998     False    False    False       False  False    False    R   
3249     False    False    False       False  False    False    L   
4371     False    False    False       False  False    False    R   

                timestamp       Date      Time  day  month  year  
5069  2019-03-04 06:54:38 2019-03-04  06:54:38    4      3  2019  
409   2016-03-30 11:54:07 2016-03-30  11:54:07   30      3  2016  
4998  2018-11-18 13:59:50 2018-11-18  13:59:50   18     11  2018  
3249  2019-11-08 22:11:00 2019-11-08  22:11:00    8     11  2019  
4371  2019-12-15 19:09:00 2019-12-15  19:09:00   15     12  2019

In [32]:
print(X_train.shape)
print(y_train.shape)
print(X_val.shape)
print(y_val.shape)

(5125, 19)
(5125,)
(1282, 19)
(1282,)


In [33]:

from sklearn.preprocessing import OneHotEncoder
import category_encoders as ce

In [34]:
encoder = ce.OneHotEncoder(cols=["year","month","day","Side"])

X_train=encoder.fit_transform(X_train)

X_val=encoder.fit_transform(X_val)

E:\Anacoda\New folder\lib\site-packages\category_encoders\utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):
E:\Anacoda\New folder\lib\site-packages\category_encoders\utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):


In [35]:
X_train.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
454   37.740299 -122.407848  False         0.000     False     False   
3198  37.709020 -122.467770  False         0.000     False     False   
4652  37.732292 -122.414909  False         0.010     False     False   
1988  37.807710 -122.367640  False         0.037     False     False   
1029  37.709324 -122.467308  False         0.000     False     False   

      Junction  No_Exit  Railway  Roundabout  ...  month_8  month_9  month_10  \
454      False    False    False       False  ...        0        0         0   
3198     False    False    False       False  ...        0        0         0   
4652      True    False    False       False  ...        0        0         0   
1988     False    False    False       False  ...        0        0         0   
1029     False    False    False       False  ...        0        0         0   

      month_11 month_12 year_1 year_2  year_3  year_4  year_5  
454          0        0      1      0       0       0       0  
3198         0        0      0      1       0       0       0  
4652         0        0      0      0       1       0       0  
1988         0        0      0      0       0       1       0  
1029         0        0      0      1       0       0       0  

[5 rows x 65 columns]

In [36]:
X_val.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
5069  37.768948 -122.406372  False           0.0     False     False   
409   37.785835 -122.391518  False           0.0     False     False   
4998  37.710304 -122.465553  False           0.0     False     False   
3249  37.770178 -122.421162  False           0.0     False     False   
4371  37.729990 -122.403729  False           0.0     False     False   

      Junction  No_Exit  Railway  Roundabout  ...  month_8  month_9  month_10  \
5069     False    False    False       False  ...        0        0         0   
409       True    False    False       False  ...        0        0         0   
4998     False    False    False       False  ...        0        0         0   
3249     False    False    False       False  ...        0        0         0   
4371     False    False    False       False  ...        0        0         0   

      month_11 month_12 year_1 year_2  year_3  year_4  year_5  
5069         0        0      1      0       0       0       0  
409          0        0      0      1       0       0       0  
4998         0        0      0      0       1       0       0  
3249         0        0      1      0       0       0       0  
4371         0        0      1      0       0       0       0  

[5 rows x 65 columns]

In [37]:
print(X_train.shape)
print(X_val.shape)

(5125, 65)
(1282, 65)


In [38]:
print(X_train.info())
print(X_val.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5125 entries, 454 to 5043
Data columns (total 65 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Lat           5125 non-null   float64       
 1   Lng           5125 non-null   float64       
 2   Bump          5125 non-null   bool          
 3   Distance(mi)  5125 non-null   float64       
 4   Crossing      5125 non-null   bool          
 5   Give_Way      5125 non-null   bool          
 6   Junction      5125 non-null   bool          
 7   No_Exit       5125 non-null   bool          
 8   Railway       5125 non-null   bool          
 9   Roundabout    5125 non-null   bool          
 10  Stop          5125 non-null   bool          
 11  Amenity       5125 non-null   bool          
 12  Side_1        5125 non-null   int64         
 13  Side_2        5125 non-null   int64         
 14  timestamp     5125 non-null   object        
 15  Date          5125 non-null   dateti

In [39]:
x_train_dropped = X_train.drop(columns=["Bump","Crossing","Give_Way","Junction","No_Exit","Railway","Roundabout","Stop","Amenity","timestamp","Time","Date"])
x_train_dropped.head()


Lat         Lng  Distance(mi)  Side_1  Side_2  day_1  day_2  \
454   37.740299 -122.407848         0.000       1       0      1      0   
3198  37.709020 -122.467770         0.000       1       0      0      1   
4652  37.732292 -122.414909         0.010       1       0      0      0   
1988  37.807710 -122.367640         0.037       1       0      0      0   
1029  37.709324 -122.467308         0.000       1       0      0      0   

      day_3  day_4  day_5  ...  month_8  month_9  month_10  month_11  \
454       0      0      0  ...        0        0         0         0   
3198      0      0      0  ...        0        0         0         0   
4652      1      0      0  ...        0        0         0         0   
1988      0      1      0  ...        0        0         0         0   
1029      0      0      1  ...        0        0         0         0   

      month_12  year_1  year_2  year_3  year_4  year_5  
454          0       1       0       0       0       0  
3198         0       0       1       0       0       0  
4652         0       0       0       1       0       0  
1988         0       0       0       0       1       0  
1029         0       0       1       0       0       0  

[5 rows x 53 columns]

In [41]:
x_val_dropped = X_val.drop(columns=["Bump","Crossing","Give_Way","Junction","No_Exit","Railway","Roundabout","Stop","Amenity","timestamp","Time","Date"])


In [42]:
#X_val.drop(columns=["Bump","Crossing","Give_Way","Junction","No_Exit","Roundabout","Stop","Amenity","timestamp","Time","Date"],inplace=True)
x_val_dropped.head()

Lat         Lng  Distance(mi)  Side_1  Side_2  day_1  day_2  \
5069  37.768948 -122.406372           0.0       1       0      1      0   
409   37.785835 -122.391518           0.0       1       0      0      1   
4998  37.710304 -122.465553           0.0       1       0      0      0   
3249  37.770178 -122.421162           0.0       0       1      0      0   
4371  37.729990 -122.403729           0.0       1       0      0      0   

      day_3  day_4  day_5  ...  month_8  month_9  month_10  month_11  \
5069      0      0      0  ...        0        0         0         0   
409       0      0      0  ...        0        0         0         0   
4998      1      0      0  ...        0        0         0         0   
3249      0      1      0  ...        0        0         0         0   
4371      0      0      1  ...        0        0         0         0   

      month_12  year_1  year_2  year_3  year_4  year_5  
5069         0       1       0       0       0       0  
409          0       0       1       0       0       0  
4998         0       0       0       1       0       0  
3249         0       1       0       0       0       0  
4371         0       1       0       0       0       0  

[5 rows x 53 columns]

In [43]:
print(X_train.info())
print(X_val.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5125 entries, 454 to 5043
Data columns (total 65 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Lat           5125 non-null   float64       
 1   Lng           5125 non-null   float64       
 2   Bump          5125 non-null   bool          
 3   Distance(mi)  5125 non-null   float64       
 4   Crossing      5125 non-null   bool          
 5   Give_Way      5125 non-null   bool          
 6   Junction      5125 non-null   bool          
 7   No_Exit       5125 non-null   bool          
 8   Railway       5125 non-null   bool          
 9   Roundabout    5125 non-null   bool          
 10  Stop          5125 non-null   bool          
 11  Amenity       5125 non-null   bool          
 12  Side_1        5125 non-null   int64         
 13  Side_2        5125 non-null   int64         
 14  timestamp     5125 non-null   object        
 15  Date          5125 non-null   dateti

In [44]:
# This cell is used to select the numerical features. IT SHOULD BE REMOVED AS YOU DO YOUR WORK.
X_train = X_train[['Lat', 'Lng', 'Distance(mi)','Side_1','Side_2','month_1','month_2','month_3','month_4','month_5','month_6','month_7','month_8','month_9','month_10','month_11','month_12']]
X_val = X_val[['Lat', 'Lng', 'Distance(mi)','Side_1','Side_2','month_1','month_2','month_3','month_4','month_5','month_6','month_7','month_8','month_9','month_10','month_11','month_12']]

In [45]:
#X_train = X_train[['Lat', 'Lng', 'Distance(mi)']]
#X_val = X_val[['Lat', 'Lng', 'Distance(mi)']]                   

In [48]:
clf=RandomForestClassifier(n_estimators=100)
clf.fit(X_train, y_train)
y_pred=clf.predict(X_val)
from sklearn import metrics
print("Accuracy:",metrics.accuracy_score(y_val, y_pred))

Accuracy: 0.8003120124804992


## Model Training

Let's train a model with the data! We'll train a Random Forest Classifier to demonstrate the process of making submissions. 

In [47]:
from sklearn.ensemble import RandomForestClassifier

# Create an instance of the classifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)

# Train the classifier
classifier = classifier.fit(X_train, y_train)

Now let's test our classifier on the validation dataset and see the accuracy.

In [49]:
print("The accuracy of the classifier on the validation set is ", (classifier.score(X_val, y_val)))

The accuracy of the classifier on the validation set is  0.7410296411856474


Well. That's a good start, right? A classifier that predicts all examples' `Severity` as 2 will get around 0.63. You should get better score as you add more features and do better data preprocessing.

## Submission File Generation

We have built a model and we'd like to submit our predictions on the test set! In order to do that, we'll load the test set, predict the class and save the submission file. 

First, we'll load the data.

In [50]:
test_df = pd.read_csv( 'test.csv')
test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp  
0  2016-04-04 19:20:31  
1  2020-10-28 11:51:00  
2  2019-09-09 07:36:45  
3  2019-08-06 15:46:25  
4  2018-10-17 09:54:58

In [51]:
test_df['Date'] = pd.to_datetime(test_df['timestamp']).dt.date
test_df['Time'] = pd.to_datetime(test_df['timestamp']).dt.time
test_df["Date"]= pd.to_datetime(test_df["Date"])
test_df['day'] = test_df['Date'].dt.day
test_df['month'] = test_df['Date'].dt.month
test_df['year'] = test_df['Date'].dt.year
test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  Roundabout   Stop  Amenity Side  \
0      True    False    False       False  False    False    R   
1     False    False    False       False  False    False    R   
2     False    False    False       False  False    False    R   
3      True    False    False       False  False    False    R   
4     False    False    False       False  False    False    R   

             timestamp       Date      Time  day  month  year  
0  2016-04-04 19:20:31 2016-04-04  19:20:31    4      4  2016  
1  2020-10-28 11:51:00 2020-10-28  11:51:00   28     10  2020  
2  2019-09-09 07:36:45 2019-09-09  07:36:45    9      9  2019  
3  2019-08-06 15:46:25 2019-08-06  15:46:25    6      8  2019  
4  2018-10-17 09:54:58 2018-10-17  09:54:58   17     10  2018

In [52]:
X_test = test_df.drop(columns=['ID'])
X_test.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  Junction  \
0  37.786060 -122.390900  False         0.039     False     False      True   
1  37.769609 -122.415057  False         0.202     False     False     False   
2  37.807495 -122.476021  False         0.000     False     False     False   
3  37.761818 -122.405869  False         0.000     False     False      True   
4  37.732350 -122.414100  False         0.670     False     False     False   

   No_Exit  Railway  Roundabout   Stop  Amenity Side            timestamp  \
0    False    False       False  False    False    R  2016-04-04 19:20:31   
1    False    False       False  False    False    R  2020-10-28 11:51:00   
2    False    False       False  False    False    R  2019-09-09 07:36:45   
3    False    False       False  False    False    R  2019-08-06 15:46:25   
4    False    False       False  False    False    R  2018-10-17 09:54:58   

        Date      Time  day  month  year  
0 2016-04-04  19:20:31    4      4  2016  
1 2020-10-28  11:51:00   28     10  2020  
2 2019-09-09  07:36:45    9      9  2019  
3 2019-08-06  15:46:25    6      8  2019  
4 2018-10-17  09:54:58   17     10  2018

In [53]:
encoder = ce.OneHotEncoder(cols=["year","month","day","Side"])

X_test=encoder.fit_transform(X_test)



E:\Anacoda\New folder\lib\site-packages\category_encoders\utils.py:21: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  elif pd.api.types.is_categorical(cols):


In [54]:
X_test.head()

Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  Junction  \
0  37.786060 -122.390900  False         0.039     False     False      True   
1  37.769609 -122.415057  False         0.202     False     False     False   
2  37.807495 -122.476021  False         0.000     False     False     False   
3  37.761818 -122.405869  False         0.000     False     False      True   
4  37.732350 -122.414100  False         0.670     False     False     False   

   No_Exit  Railway  Roundabout  ...  month_8  month_9  month_10  month_11  \
0    False    False       False  ...        0        0         0         0   
1    False    False       False  ...        0        0         0         0   
2    False    False       False  ...        0        0         0         0   
3    False    False       False  ...        0        0         0         0   
4    False    False       False  ...        0        0         0         0   

  month_12 year_1 year_2  year_3  year_4  year_5  
0        0      1      0       0       0       0  
1        0      0      1       0       0       0  
2        0      0      0       1       0       0  
3        0      0      0       1       0       0  
4        0      0      0       0       1       0  

[5 rows x 65 columns]

Note that the test set has the same features and doesn't have the `Severity` column.
At this stage one must **NOT** forget to apply the same processing done on the training set on the features of the test set.

Now we'll add `Severity` column to the test `DataFrame` and add the values of the predicted class to it.

**I'll select the numerical features here as I did in the training set. DO NOT forget to change this step as you change the preprocessing of the training data.**

In [55]:


# You should update/remove the next line once you change the features used for training
X_test = X_test[['Lat', 'Lng', 'Distance(mi)','Side_1','Side_2','month_1','month_2','month_3','month_4','month_5','month_6','month_7','month_8','month_9','month_10','month_11','month_12']]

y_test_predicted = classifier.predict(X_test)

test_df['Severity'] = y_test_predicted

test_df.head()

ID        Lat         Lng   Bump  Distance(mi)  Crossing  Give_Way  \
0  6407  37.786060 -122.390900  False         0.039     False     False   
1  6408  37.769609 -122.415057  False         0.202     False     False   
2  6409  37.807495 -122.476021  False         0.000     False     False   
3  6410  37.761818 -122.405869  False         0.000     False     False   
4  6411  37.732350 -122.414100  False         0.670     False     False   

   Junction  No_Exit  Railway  ...   Stop  Amenity  Side            timestamp  \
0      True    False    False  ...  False    False     R  2016-04-04 19:20:31   
1     False    False    False  ...  False    False     R  2020-10-28 11:51:00   
2     False    False    False  ...  False    False     R  2019-09-09 07:36:45   
3      True    False    False  ...  False    False     R  2019-08-06 15:46:25   
4     False    False    False  ...  False    False     R  2018-10-17 09:54:58   

        Date      Time day  month  year  Severity  
0 2016-04-04  19:20:31   4      4  2016         2  
1 2020-10-28  11:51:00  28     10  2020         2  
2 2019-09-09  07:36:45   9      9  2019         2  
3 2019-08-06  15:46:25   6      8  2019         2  
4 2018-10-17  09:54:58  17     10  2018         2  

[5 rows x 21 columns]

Now we're ready to generate the submission file. The submission file needs the columns `ID` and `Severity` only.

In [56]:
test_df[['ID', 'Severity']].to_csv('submission.csv', index=False)

The remaining steps is to submit the generated file and are as follows. 

1. Press `Save Version` on the upper right corner of this notebook.
2. Write a `Version Name` of your choice and choose `Save & Run All (Commit)` then click `Save`.
3. Wait for the saved notebook to finish running the go to the saved notebook.
4. Scroll down until you see the output files then select the `submission.csv` file and click `Submit`.

Now your submission will be evaluated and your score will be updated on the leaderboard! CONGRATULATIONS!!

## Conclusion

In this notebook, we have demonstrated the essential steps that one should do in order to get "slightly" familiar with the data and the submission process. We chose not to go into details in each step to keep the welcoming notebook simple and make a room for improvement.

You're encourged to `Fork` the notebook, edit it, add your insights and use it to create your submission.